syntax = "proto3";

package task;

option go_package = "github.com/indoor-mall-nav/navign/admin/tower/proto/task";

// TaskType defines the type of task to be executed
enum TaskType {
  TASK_TYPE_UNSPECIFIED = 0;
  TASK_TYPE_DELIVERY = 1;
  TASK_TYPE_PATROL = 2;
  TASK_TYPE_RETURN_HOME = 3;
  TASK_TYPE_EMERGENCY = 4;
}

// Priority level for task execution
enum Priority {
  PRIORITY_UNSPECIFIED = 0;
  PRIORITY_LOW = 1;
  PRIORITY_NORMAL = 2;
  PRIORITY_HIGH = 3;
  PRIORITY_URGENT = 4;
}

// Location represents a point with coordinates
message Location {
  double x = 1;
  double y = 2;
  double z = 3;
  string floor = 4;
}

// Task represents a work item to be assigned to a robot
message Task {
  string id = 1;
  TaskType type = 2;
  repeated Location sources = 3;  // Starting locations
  repeated Location terminals = 4;  // Destination locations
  Priority priority = 5;
  int64 created_at = 6;  // Unix timestamp
  string entity_id = 7;  // Entity/mall identifier
  map<string, string> metadata = 8;  // Additional task parameters
}

// TaskRequest is sent from the Rust orchestrator to assign a task
message TaskRequest {
  Task task = 1;
}

// TaskResponse acknowledges task receipt and assignment
message TaskResponse {
  bool accepted = 1;
  string robot_id = 2;  // ID of robot assigned to task
  string message = 3;
  int64 estimated_completion_time = 4;  // Unix timestamp
}

// RobotStatus represents the current state of a robot
enum RobotState {
  ROBOT_STATE_UNSPECIFIED = 0;
  ROBOT_STATE_IDLE = 1;
  ROBOT_STATE_BUSY = 2;
  ROBOT_STATE_CHARGING = 3;
  ROBOT_STATE_ERROR = 4;
  ROBOT_STATE_OFFLINE = 5;
}

// RobotInfo contains robot identification and status
message RobotInfo {
  string id = 1;
  string name = 2;
  RobotState state = 3;
  Location current_location = 4;
  double battery_level = 5;  // Percentage 0-100
  string current_task_id = 6;
  int64 last_seen = 7;  // Unix timestamp
  string entity_id = 8;
}

// RobotDistributionRequest asks for current robot status
message RobotDistributionRequest {
  string entity_id = 1;
}

// RobotDistributionResponse returns information about all robots
message RobotDistributionResponse {
  repeated RobotInfo robots = 1;
  int32 total_count = 2;
  int32 idle_count = 3;
  int32 busy_count = 4;
}

// RobotReportRequest is sent from Go tower to Rust orchestrator
message RobotReportRequest {
  RobotInfo robot = 1;
}

// RobotReportResponse acknowledges the report
message RobotReportResponse {
  bool success = 1;
  string message = 2;
}

// TaskAssignment is sent from Rust orchestrator to Go tower
message TaskAssignment {
  string robot_id = 1;
  Task task = 2;
}

// TaskAssignmentResponse acknowledges task assignment
message TaskAssignmentResponse {
  bool accepted = 1;
  string message = 2;
}

// ============================================================================
// Plot Extraction Messages (for floor plan polygon extraction)
// ============================================================================

// 2D Point for polygon vertices
message Point {
  double x = 1;
  double y = 2;
}

// Polygon representing an extracted area from floor plan
message Polygon {
  repeated Point vertices = 1;
  string label = 2;  // Optional label/identifier
  double area = 3;  // Area in square pixels
  Point centroid = 4;  // Center point
}

// Image data for floor plans
message FloorPlanImage {
  bytes data = 1;
  string format = 2;  // e.g., "png", "jpeg"
  int32 width = 3;
  int32 height = 4;
}

// Configuration for polygon extraction algorithm
message PlotExtractionConfig {
  // Preprocessing
  double blur_kernel_size = 1;  // Gaussian blur kernel (default: 5.0)
  int32 threshold_value = 2;  // Binary threshold (0-255, default: 127)
  int32 threshold_type = 3;  // 0=BINARY, 1=OTSU (default: 0)

  // Contour filtering
  double min_area = 4;  // Minimum polygon area (default: 100.0)
  double max_area = 5;  // Maximum polygon area (0 = no limit)
  double epsilon_factor = 6;  // Contour approximation (default: 0.01)

  // Morphological operations
  bool apply_morphology = 7;  // Apply morphology (default: true)
  int32 morph_kernel_size = 8;  // Kernel size (default: 5)

  // Edge detection (optional)
  bool use_canny = 9;  // Use Canny edge detection (default: false)
  int32 canny_low = 10;  // Canny low threshold (default: 50)
  int32 canny_high = 11;  // Canny high threshold (default: 150)
}

// Request to extract polygons from floor plan
message ExtractPolygonsRequest {
  string entity_id = 1;
  string floor_id = 2;
  FloorPlanImage floor_plan = 3;
  PlotExtractionConfig config = 4;  // Optional, uses defaults if not provided
}

// Response with extracted polygons
message ExtractPolygonsResponse {
  repeated Polygon polygons = 1;
  int32 total_count = 2;
  string error = 3;  // Error message if extraction failed
  PlotProcessingStats stats = 4;
}

// Processing statistics
message PlotProcessingStats {
  int32 contours_found = 1;
  int32 contours_filtered = 2;
  double processing_time_ms = 3;
  int32 image_width = 4;
  int32 image_height = 5;
}

// Request for batch processing
message BatchExtractRequest {
  string entity_id = 1;
  repeated FloorPlanBatch floor_plans = 2;
  PlotExtractionConfig config = 3;
}

// Single floor plan in batch
message FloorPlanBatch {
  string floor_id = 1;
  FloorPlanImage floor_plan = 2;
}

// Response for batch processing
message BatchExtractResponse {
  repeated FloorExtraction extractions = 1;
  int32 successful = 2;
  int32 failed = 3;
}

// Single floor extraction result
message FloorExtraction {
  string floor_id = 1;
  repeated Polygon polygons = 2;
  string error = 3;
  PlotProcessingStats stats = 4;
}

// ============================================================================
// Services
// ============================================================================

// OrchestratorService is implemented by Rust orchestrator (server)
// Go tower acts as client to report robot status and receive task assignments
service OrchestratorService {
  // ReportRobotStatus is called by Go tower to report robot status
  rpc ReportRobotStatus(RobotReportRequest) returns (RobotReportResponse);

  // GetTaskAssignment is called by Go tower to get tasks for robots
  // This could be streaming or polling based on implementation
  rpc GetTaskAssignment(RobotDistributionRequest) returns (stream TaskAssignment);

  // Plot extraction methods (called by Python plot client)

  // ExtractPolygons extracts polygons from a single floor plan
  // Python plot client calls this with the floor plan image
  rpc ExtractPolygons(ExtractPolygonsRequest) returns (ExtractPolygonsResponse);

  // BatchExtractPolygons extracts polygons from multiple floor plans
  rpc BatchExtractPolygons(BatchExtractRequest) returns (BatchExtractResponse);
}
